app/controllers/uploads_controller.rb

```diff
@@ -270,6 +301,186 @@ def complete_external_upload
     end
   end
 
+  def create_multipart
+    return render_404 if !SiteSetting.enable_direct_s3_uploads
+
+    RateLimiter.new(
+      current_user, "create-multipart-upload", CREATE_MULTIPART_RATE_LIMIT_PER_MINUTE, 1.minute
+    ).performed!
+
+    file_name = params.require(:file_name)
+    file_size = params.require(:file_size).to_i
+    upload_type = params.require(:upload_type)
+    content_type = params.require(:content_type)
+
+    if file_size_too_big?(file_name, file_size)
+      return render_json_error(
+        I18n.t("upload.attachments.too_large", max_size_kb: SiteSetting.max_attachment_size_kb),
+        status: 422
+      )
+    end
+
+    begin
+      multipart_upload = Discourse.store.create_multipart(
+        file_name, content_type
+      )
+    rescue Aws::S3::Errors::ServiceError => err
+      debug_upload_error(err, "upload.create_mutlipart_failure")
+      return render_json_error(I18n.t("upload.failed"), status: 422)
+    end
+
+    upload_stub = ExternalUploadStub.create!(
+      key: multipart_upload[:key],
+      created_by: current_user,
+      original_filename: file_name,
+      upload_type: upload_type,
+      external_upload_identifier: multipart_upload[:upload_id],
+      multipart: true,
+      filesize: file_size
+    )
+
+    render json: {
+      external_upload_identifier: upload_stub.external_upload_identifier,
+      key: upload_stub.key,
+      unique_identifier: upload_stub.unique_identifier
+    }
+  end
+
+  def batch_presign_multipart_parts
+    return render_404 if !SiteSetting.enable_direct_s3_uploads
+
+    part_numbers = params.require(:part_numbers)
+    unique_identifier = params.require(:unique_identifier)
+
+    RateLimiter.new(
+      current_user, "batch-presign-#{unique_identifier}", BATCH_PRESIGN_RATE_LIMIT_PER_MINUTE, 1.minute
```

Comment:
Since the `unique_identifier` is from a param, this rate limit can easily be bypassed by passing a different unique identifier manually so I'm not sure if this rate limit will serve its purpose.

Reply:
Hmmm yes I think I can just rate limit without the `unique_identifier`